### PR TITLE
Implement SharesAPI with in memory persistence

### PR DESCRIPTION
**Pull request by duhizjame**

### CODE REVIEW

**Review by matar993 — CHANGES_REQUESTED** [review]

File: server/src/test/java/io/delta/sharing/api/server/DeltaSharesApiImplTest.java [hunk 1]
```diff
@@ -18,4 +22,15 @@ public void getUnknownShare() {
...
+  @Test
+  public void listShares() throws IOException {
+    given()
+        .queryParam("maxResults", 50.0)
+        .queryParam("pageToken", encoder.encodePageToken("0"))
+        .when()
+        .get("/shares")
+        .then()
+        .statusCode(200);
+  }
```
Comment (matar993) [inline server/src/test/java/io/delta/sharing/api/server/DeltaSharesApiImplTest.java:35]: A test without `maxResults` and `pageToken` query params is needed.

File: server/src/main/java/io/delta/sharing/api/server/DeltaSharesApiImpl.java [hunk 2]
```diff
@@ -79,8 +84,19 @@ public CompletionStage<Response> listSchemas(
...
   @Override
   public CompletionStage<Response> listShares(BigDecimal maxResults, String pageToken) {
-    Response res = Response.ok().build();
-    return CompletableFuture.completedFuture(res);
+    return CompletableFuture.supplyAsync(
+        () -> {
+          try {
+            ImmutablePair<List<Share>, Optional<String>> shares =
+                deltaSharesService.listShares(
+                    Optional.of(pageToken), Optional.of(maxResults.intValue()));
```
[the remark this answers is not in the document]
Comment (matar993) [reply server/src/main/java/io/delta/sharing/api/server/DeltaSharesApiImpl.java:92]: Also you must use `Optional.ofNullable()`

**Comment by duhizjame** [comment]
@matar993 resolved comments via the latest commit

File: server/src/main/java/io/delta/sharing/api/server/PagingUtils.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package io.delta.sharing.api.server;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class PagingUtils {
```
Comment (tmnd1991) [inline server/src/main/java/io/delta/sharing/api/server/PagingUtils.java:10]: Remove this

**Review by tmnd1991 — APPROVED** [review]

**Pull request merged**